sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/parquet/ParquetFilters.scala

```diff
@@ -597,9 +597,9 @@ class ParquetFilters(
         createFilterHelper(pred, canPartialPushDownConjuncts = false)
           .map(FilterApi.not)
 
-      case sources.In(name, values) if canMakeFilterOn(name, values.head)
-        && values.distinct.length <= pushDownInFilterThreshold =>
-        values.distinct.flatMap { v =>
```

Comment:
Hm, it's better to don't rely on optimizer actually. 5000000 N in the benchmark you did looks a bit unrealistic and the perf degradation seems not very serious compared to execution time.

Reply:
+1 with @HyukjinKwon 
@wangyum I think this PR can cause perf regression on filter pushdown in Parquet. After the changes, `In` filters with redundant elements might not be able to be pushed down.

Reply:
@HyukjinKwon @gengliangwang If we do not rely on the optimizer, we should add an empty check. otherwise `values.head` will throw an exception.